disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/exporters/AbstractDiskExporter.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.disk.buffering.internal.exporters;
+
+import io.opentelemetry.contrib.disk.buffering.internal.serialization.serializers.SignalSerializer;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.FolderManager;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.Storage;
+import io.opentelemetry.contrib.disk.buffering.storage.StorageConfiguration;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class AbstractDiskExporter<EXPORT_DATA> {
+  private final Storage storage;
+  private static final Logger logger = Logger.getLogger(AbstractDiskExporter.class.getName());
+
+  public AbstractDiskExporter(File rootDir, StorageConfiguration configuration) {
+    validateConfiguration(configuration);
+    this.storage = new Storage(new FolderManager(getSignalFolder(rootDir), configuration));
+  }
+
+  /**
+   * Reads data from the disk and attempts to export it.
+   *
+   * @param timeout The amount of time to wait for the wrapped exporter to finish.
+   * @param unit The unit of the time provided.
+   * @return TRUE if there was data available and it was successfully exported within the timeout
+   *     provided. FALSE if either of those conditions didn't meet.
+   * @throws IOException If an unexpected error happens.
+   */
+  public boolean exportStoredBatch(long timeout, TimeUnit unit) throws IOException {
+    logger.log(Level.INFO, "Attempting to export batch from disk.");
+    AtomicBoolean exportSucceeded = new AtomicBoolean(false);
+    boolean foundDataToExport =
+        storage.read(
```

Comment:
I don't understand what happens if the export thread gets killed after `storage.read`? Do we assume that the exporter thread lives forever?

I see that `ReadableFile` has an index for the current reading position, but I can't see an index of the current file that is being read - `findReadableFile` seems to pick the first stored file.

Maybe add a CONTRIBUTING.md file with a simple data flow diagram (e.g. with https://asciiflow.com) would help.

Reply:
I'm not sure I'm following the question. Though it seems like the concern is related to how to continue reading from the same file after the app gets destroyed and reopened?

`findReadableFile` will always get the oldest (unexpired) available file, so if the reading process gets killed after starting to read from a file, what would happen when a new read process is started is:

- findReadableFile will search for the oldest available file, which should be the same file it was being read before, so it will pick up where it left off.
- If the oldest file, which is the same as it was being read before, has expired due to the configured [maxFileAgeForRead](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/913/files/549ea97609ea7ee71ae2e069fd1e8cfb4ea2c9da#diff-f2b3e0dc8ee13418cbbdc79a8d73e8c201fae10caea943142626dd8c448c74a2R30) parameter, then the leftover data in that file will be discarded because it's stale, and the next oldest file will be picked up.

I hope it helps clarify it. And yeah I agree a data flow diagram would be really helpful. I'll find the time to create after the code changes are done.

Reply:
That's correct, it doesn't matter what was sent before when it comes to picking what file to read from, as long as the file has not expired. Now, if the same file is chosen more than once, the second time it is read it won't provide the data that was previously sent though, this is because the data is removed from the file as soon as it gets successfully exported.

Reply:
> because the data is removed from the file as soon as it gets successfully exported.

where is that logic?

Reply:
It's [in here](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/913/files#diff-fb0920b798b73be9fdbda266cabf3752079478b8d8ac91983c3ee11b754da5fcR86).